### ISSUE TITLE
Re-Engineer Namespaces

### ISSUE THREAD

**Issue by magnus-madsen** [flix/flix#4746]
The purpose of this ticket is to streamline the namespace mechanism. 

The current namespace mechanism is a bit idiosyncratic and has some strange corner cases.

By carefully refactoring, inspired by Rust's namespace mechanism, I think we can arrive in a great place.

We need these (breaking) changes:

- [ ] Declarations in a parent namespace are not visible in a sub namespace (but must be explicitly imported).
  - Detail: The root namespace is always visible (e.g. Prelude).
- [ ] Opening a new namespace will clear all `use`s and `import`s. 
- [ ] A name cannot refer to more than one entity:
  - [ ] It is illegal to shadow a declaration or type with a `use` or `import`.
  - [ ] It is illegal to shadow a `use` or `import` with a declaration or type.
- [ ] We unify the namespace of Flix types and Java types (i.e. uenv and ienv should ultimately be merged). 

I believe these changes can be implemented one-at-a-time with some changes to the standard library required.

**Comment by JonathanStarup** [comment]
A question: If i have two files with the same namespace, what is the scope of imports and uses? Does the below example compile?
```
// File 1
namespace B { pub def function(x: Int32): Int32 = x }
namespace A { use B.function }
```
```
// File 2
namespace A {
    pub def example(x: Int32): Int32 = function(x)
}
```

**Comment by JonathanStarup** [comment]
Another question: Now that uses/imports are namespace local, what does it mean to use/import something at the top level?

**Comment by magnus-madsen** [comment]
> Another question: Now that uses/imports are namespace local, what does it mean to use/import something at the top level?

It means they are only visible at the top-level. In other words, it would only be useful for small programs.

**Comment by magnus-madsen** [comment]
> A question: If i have two files with the same namespace, what is the scope of imports and uses? Does the below example compile?

Imports and uses are per file (like currently), hence your program does not compile.

In the future we may disallow declaring the same namespace twice, and then it would even more not compile.

**Comment by JonathanStarup** [comment]
> > Another question: Now that uses/imports are namespace local, what does it mean to use/import something at the top level?
> 
> It means they are only visible at the top-level. In other words, it would only be useful for small programs.

Top level imports could also mean, import this for the whole file, no matter the namespace. This would be nice for very general imports like `ToString.toString`

**Comment by magnus-madsen** [comment]
If you think about a namespace as more of a module then the nice thing about (having to write):

```
namespace Foo {
    use Bar.XYZ;
    use Baz.XYZ;
}
```

is that you can see `Bar.XYZ` and `Baz.XYZ` as the "dependencies" that `Foo` depends on. Its almost like arguments to a constructor.

**Comment by JonathanStarup** [comment]
I think that conceptualization still holds with file-wide `use`, just the syntax is different, but I see your point. I think I've resigned myself to the opinion that namespaces and imports are just tedious, no matter what, without the help of the editor (We still need namespace aliasing, of course)

**Comment by paulbutcher** [comment]
Questions:

Do we still intend to support declaring namespaces more than once? In which case, is this legal?

```
namespace Foo {
  use Bar.XYZ
}

namespace Bar {
  enum XYZ
}

namespace Foo {
  def foo(): XYZ = ???
}
```

**Comment by paulbutcher** [comment]
And similarly, this is illegal?

```
namespace Foo {
  use Bar.XYZ
}

namespace Bar {
  enum XYZ
}

namespace Foo {
  enum XYZ
}
```

**Comment by magnus-madsen** [comment]
> Questions:
> 
> Do we still intend to support declaring namespaces more than once? In which case, is this legal?
> 
> ```
> namespace Foo {
>   use Bar.XYZ
> }
> 
> namespace Bar {
>   enum XYZ
> }
> 
> namespace Foo {
>   def foo(): XYZ = ???
> }
> ```

I think initially yes, but in the next iteration no. I would like to couple the "no re+declaration of namespaces" with another change where by you can declare namespaces by placing them into a specific files (following Rust). 

Also because I would like to make these changes incrementally.

**Comment by magnus-madsen** [comment]
Oh, also, we rely on "re-declared" namespaces in some cases and we should resolve those first.

**Comment by paulbutcher** [comment]
> Also because I would like to make these changes incrementally.

This is why I'm asking the question. Right now, the machinery for `use` and `import` (i.e. `UseEnv` and `ImportEnv` in the namer) make this very difficult.

If we're going to eventually disallow redeclaration of namespaces (which I don't see in the list of changes above?) then I suggest that should be one of the first things we do: it will make rationalising `UseEnv` and `ImportEnv` much easier if there's only one instance of a namespace.

**Comment by JonathanStarup** [comment]
Another thing to consider: should you be able to re-export `use`s?

**Comment by paulbutcher** [comment]
To make this concrete: the end-game is, I think, that imports/uses should eventually work just like all other names: i.e. they should be members of `NamedAst.Root` rather than separate environments which are only maintained within the Namer.

(a peripheral benefit of which is that it will make the argument lists in the namer, which are getting out of hand, shorter).

This will be much easier to implement if we disallow re-opening namespaces.

**Comment by magnus-madsen** [comment]
> > Also because I would like to make these changes incrementally.
> 
> This is why I'm asking the question. Right now, the machinery for `use` and `import` (i.e. `UseEnv` and `ImportEnv` in the namer) make this very difficult.
> 
> If we're going to eventually disallow redeclaration of namespaces (which I don't see in the list of changes above?) then I suggest that should be one of the first things we do: it will make rationalising `UseEnv` and `ImportEnv` much easier if there's only one instance of a namespace.

Are you sure? I think we process each namespace separately, hence they already have separate uses/imports?

**Comment by magnus-madsen** [comment]
> Another thing to consider: should you be able to re-export `use`s?

No. At least not now.

**Comment by paulbutcher** [comment]
> Another thing to consider: should you be able to re-export `use`s?

Good question. My instinct is eventually yes, if they're going to become "first class" members of a namespace. Although not necessarily immediately.

**Comment by paulbutcher** [comment]
> Are you sure? I think we process each namespace separately, hence they already have separate uses/imports?

That's certainly true right now. Which means that making this legal is currently very difficult:

```
namespace Foo {
  use Bar.XYZ
}

namespace Bar {
  enum XYZ
}

namespace Foo {
  def foo(): XYZ = ???
}
```

**Comment by paulbutcher** [comment]
And similarly making this illegal is currently very difficult:

```
namespace Foo {
  use Bar.XYZ
}

namespace Bar {
  enum XYZ
}

namespace Foo {
  enum XYZ
}
```

**Comment by magnus-madsen** [comment]
```
namespace Foo {
  use Bar.XYZ
}

namespace Bar {
  enum XYZ
}

namespace Foo {
  def foo(): XYZ = ???
}
```

This should not be legal.

**Comment by paulbutcher** [comment]
> This should not be legal.

I'm confused. Above you said:

> I think initially yes, but in the next iteration no.

**Comment by magnus-madsen** [comment]
First, we agree on the ultimate goal, I just want to get there by small steps.

In this program:
```
namespace Foo {
  use Bar.XYZ
}

namespace Bar {
  enum XYZ
}

namespace Foo {
  def foo(): XYZ = ???
}
```

It is OK that `Foo` is opened twice. It is not OK to refer to XYZ inside the second declaration because XYZ is not in scope.

**Comment by magnus-madsen** [comment]
If we do it this way, I believe the changes can be done incrementally while things make sense.

**Comment by paulbutcher** [comment]
Ah, OK. So we've been talking at cross-purposes.

In which case, I agree that the order doesn't make any difference.

Do you agree with:

> To make this concrete: the end-game is, I think, that imports/uses should eventually work just like all other names: i.e. they should be members of NamedAst.Root rather than separate environments which are only maintained within the Namer.
> 
> (a peripheral benefit of which is that it will make the argument lists in the namer, which are getting out of hand, shorter).

**Comment by magnus-madsen** [comment]
> To make this concrete: the end-game is, I think, that imports/uses should eventually work just like all other names: i.e. they should be members of NamedAst.Root rather than separate environments which are only maintained within the Namer.

I am not sure what this means.

Do you mean before or after name resolution?

**Comment by paulbutcher** [comment]
I mean that we should (eventually) treat names introduced by `use` and `import` in exactly the same way as we treat all other names. Right now, we maintain *almost* all names as members of `NamedAst.Root`:

```
  case class Root(classes: Map[Name.NName, Map[String, NamedAst.Class]],
                  instances: Map[Name.NName, Map[String, List[NamedAst.Instance]]],
                  defsAndSigs: Map[Name.NName, Map[String, NamedAst.DefOrSig]],
                  enums: Map[Name.NName, Map[String, NamedAst.Enum]],
                  typeAliases: Map[Name.NName, Map[String, NamedAst.TypeAlias]],
                  effects: Map[Name.NName, Map[String, NamedAst.Effect]],
                  ops: Map[Name.NName, Map[String, NamedAst.Op]],
                  entryPoint: Option[Symbol.DefnSym],
                  sources: Map[Source, SourceLocation])
```

But we treat names introduced via `use` and `import` differently: they are instead handled using `UseEnv` and `ImportEnv`, which only exist within the namer. They're currently passed to every method because they have to work recursively (which other names do not).

So ultimately, I think we should handle them just like every other name.

Obviously we should get there incrementally, not in a big bang: but I want to check that we're aligned on the endgame.

**Comment by magnus-madsen** [comment]
We could probably maintain some maps inside `Root`, but the reasons that uenv/ienv have to be parameters is that they depend on the scope. In particular, both `use` and `import` can be used as expressions (which means they have to be resolved together with local variables)-- so the need for these arguments is inherent in the problem. But maintaining some additional separate namespace information would be useful.

**Comment by paulbutcher** [comment]
>  both use and import can be used as expressions

They can?! Can you give me an example?

**Comment by magnus-madsen** [comment]
https://github.com/flix/flix/blob/master/main/src/library/String.flix#L273
https://github.com/flix/flix/blob/master/main/src/library/Fixpoint/Solver.flix#L116

**Comment by paulbutcher** [comment]
> https://github.com/flix/flix/blob/master/main/src/library/String.flix#L273
> https://github.com/flix/flix/blob/master/main/src/library/Fixpoint/Solver.flix#L116

Oh, I see what you mean. Sorry, I guess I don't think of those as expressions (because they don't really have a value: you couldn't pass them to anything, for example), although I guess they are strictly speaking expressions in the grammar...

**Comment by magnus-madsen** [comment]
I think of them as let-expressions `let x =e1; e2`. They bind a name to a value in a body expression.

**Comment by paulbutcher** [comment]
Just for reference, I thought I'd look to see how often we actually reopen namespaces. Excluding tests (which re-use the same namespace all the time, but typically not within a single test) it seems like we only do it heavily in the fixpoint code. The only other namespace that's ever re-opened is `Time`:

```
paulbutcher@Pauls-MBP flix % grep -h -R "namespace .* {" * --include "*.flix" --exclude "*Test*" | sort | uniq -c      
   1     namespace Instant {
   1     namespace StdErr {
   1     namespace StdIn {
   1     namespace StdOut {
   1 namespace Applicative {
   1 namespace Array {
   1 namespace Assert {
   1 namespace Belnap {
   1 namespace Bench/Chain {
   1 namespace Bench/Fixpoint {
   1 namespace Bench/Fusion {
   1 namespace Benchmark {
   1 namespace BenchmarkConstraint {
   1 namespace BenchmarkList {
   1 namespace BenchmarkMap {
   1 namespace BenchmarkSet {
   1 namespace BigDecimal {
   1 namespace BigInt {
   1 namespace Bool {
   1 namespace Boxable {
   1 namespace Chain {
   1 namespace Channel {
   1 namespace Char {
   1 namespace Choice {
   1 namespace Comparison {
   1 namespace Concurrent/Channel {
   1 namespace Concurrent/Condition {
   1 namespace Concurrent/ReentrantLock {
   1 namespace Console {
   1 namespace ConstantProp {
   1 namespace DelayList {
   1 namespace DelayMap {
   1 namespace Domain/Constant {
   1 namespace Domain/ConstantParity {
   1 namespace Domain/ConstantSign {
   1 namespace Domain/Interval {
   1 namespace Domain/IntervalAlt {
   1 namespace Domain/Mod3 {
   1 namespace Domain/Parity {
   1 namespace Domain/ParitySign {
   1 namespace Domain/PrefixSuffix {
   1 namespace Domain/Sign {
   1 namespace Domain/StrictSign {
   1 namespace Environment {
   1 namespace File {
  11 namespace Fixpoint {
  11 namespace Fixpoint/Ast {
   6 namespace Fixpoint/Ram {
   1 namespace Fixpoint/Shared {
   1 namespace Fixpoint/Tuple {
   1 namespace Float32 {
   1 namespace Float64 {
   1 namespace Foldable {
   1 namespace Functor {
   1 namespace GetOpt {
   1 namespace Graph {
   1 namespace Graphs {
   1 namespace Hash {
   1 namespace IDE {
   1 namespace Int16 {
   1 namespace Int32 {
   1 namespace Int64 {
   1 namespace Int8 {
   1 namespace Iterator {
   1 namespace List {
   1 namespace Map {
   1 namespace Monad {
   1 namespace Monoid {
   1 namespace MultiMap {
   1 namespace MutDeque {
   1 namespace MutList {
   1 namespace MutMap {
   1 namespace MutSet {
   1 namespace Nec {
   1 namespace Nel {
   1 namespace Object {
   1 namespace Option {
   1 namespace Order {
   1 namespace Random {
   1 namespace Reachable/Functional {
   1 namespace Reachable/Imperative {
   1 namespace Reachable/Logic {
   1 namespace RedBlackTree {
   1 namespace Result {
   1 namespace SemiGroup {
   1 namespace Set {
   1 namespace Stratifier {
   1 namespace String {
   1 namespace StringBuilder {
   1 namespace System {
   1 namespace Thread {
   2 namespace Time {
   1 namespace Time/Duration {
   1 namespace Time/Epoch {
   1 namespace Traversable {
   1 namespace Validation {
```

**Comment by magnus-madsen** [comment]
(1) Yes, this is what I would expect. (2) Note that the root namespace is often "reopened". (3) The Time thing should be easy to fix. (4) We should ask about Stephen and Matts projects.

**Comment by paulbutcher** [comment]
@stephentetley @mlutze To save you wading through all the above:

One of the changes that we're proposing is to disallow re-opening (re-declaring) a namespace. So it will be illegal to write:

```
namespace A { ... }
namespace A { ... }
```

Either within a single file, or separate files within a project.

Within the Flix code, this facility is only currently used heavily within the fixpoint code. But it's possible that you are making use of it in your own projects?

Are you? And how much of an issue would this change be if you are?

**Comment by paulbutcher** [comment]
So, going through the requirements above:

> Declarations in a parent namespace should not be visible in a sub namespace.
> Detail: The root namespace is always visible (e.g. Prelude).

This is already true, I believe? So the following does not compile, with errors saying that both `Foo` and `foo` are undefined:

```
namespace A {
    pub enum Foo

    def foo(): Foo = ???

    namespace B {
        def bar(): Foo = foo()
    }
}
```

So there's nothing to do for this requirement? Or am I missing something?

**Comment by paulbutcher** [comment]
> A name cannot refer to more than one entity:
>  It is illegal to shadow a declaration or type with a use or import.
>  It is illegal to shadow a use or import with a declaration or type.

This is also, I believe, done (this is what I implemented in #4694)?

**Comment by magnus-madsen** [comment]
I think we need to add negative test cases. That will help ensure that a specific functionality is correctly implemented. 

For example, I notice that this https://github.com/flix/flix/blob/master/main/src/ca/uwaterloo/flix/language/phase/Namer.scala#L84 does not seem to clear the available uses/imports.

**Comment by paulbutcher** [comment]
Right, but clearing uses/imports is:

> Opening a new namespace will clear all uses and imports.

Which is the only functionality that's currently missing?

**Comment by paulbutcher** [comment]
How about this: I'll write some test cases which confirm that:

> Declarations in a parent namespace should not be visible in a sub namespace.
> Detail: The root namespace is always visible (e.g. Prelude).

is already true (I think it is).

I don't need to write any test cases to confirm that:

> A name cannot refer to more than one entity:
> It is illegal to shadow a declaration or type with a use or import.
> It is illegal to shadow a use or import with a declaration or type.

is already true (I did that when implementing https://github.com/flix/flix/pull/4694).

Assuming that's all OK (which it will be) then I'll move on to implementing:

> Opening a new namespace will clear all uses and imports.

?

**Comment by mlutze** [comment]
> One of the changes that we're proposing is to disallow re-opening (re-declaring) a namespace. So it will be illegal to write:
> 
> ```
> namespace A { ... }
> namespace A { ... }
> ```
> 
> Either within a single file, or across separate files within a project.
> 
> Within the Flix code, this facility is only currently used heavily within the fixpoint code. But it's possible that you are making use of it in your own projects?
> 
> Are you? And how much of an issue would this change be if you are?

One use case I have for this is the "companion namespaces idiom"

```scala
namespace ProjectRoot {
// some stuff
}
... new file ...
namespace ProjectRoot {
    enum MyDataThingy
    namespace MyDataThingy {
          def convertMyDataThingyToCookie(...) ...
    }
}
```

This isn't terribly difficult to go through and fix, but it results in ugly redundant imports:
```scala
use ProjectRoot/MyDataThingy.MyDataThingy // import  the type
use ProjectRoot/MyDataThingy.MyDataThingy.MyDataThingy // import the singleton constructor
```
```

**Comment by stephentetley** [comment]
I use quite vanilla namespaces and haven't knowingly provided an extension to an existing namespace.
 
With the current implementation I've always assumed you could extend the stdlib in **local code**: 

~~~

namespace List {

    pub def isSingleton(l: List): Bool = match l {
        case _ :: Nil => true
        case _        => false
    }

}

~~~

This gives us something very much like F#'s Type Extensions but it does seem rather unprincipled (in F# and in Flix).

**Comment by JonathanStarup** [comment]
> One use case I have for this is the "companion namespaces idiom"

I feel like `MyDataThingy` should be in the `MyDataThingy` namespace and then have `use MyDataThingy.MyDataThingy` in `ProjectRoot`. Isn't this an easier solution that also supports opaque enums? Also conceptually sensible (the `MyDataThingy` namespace knows about the `MyDataThingy` type) in a way that doesn't need compiler magic like companions does.
If the goal is to have external code be able to write `ProjectRoot.MyDataThingy` then I'd use `export`.

**Comment by mlutze** [comment]
> > One use case I have for this is the "companion namespaces idiom"
> 
> I feel like `MyDataThingy` should be in the `MyDataThingy` namespace and then have `use MyDataThingy.MyDataThingy` in `ProjectRoot`. Isn't this an easier solution that also supports opaque enums? Also conceptually sensible (the `MyDataThingy` namespace knows about the `MyDataThingy` type) in a way that doesn't need compiler magic like companions does. If the goal is to have external code be able to write `ProjectRoot.MyDataThingy` then I'd use `export`.
> 
> It feels like this pattern tries to emulate classes with static methods but I think its a syntax problem that can be solved with one import and change of pattern instead of compiler support.

My concern about re-exporting is that now I have `ProjectRoot.MyDataThingy` and `ProjectRoot.MyDataThingy.MyDataThingy` and they both mean the same thing. This kind of thing has confused me when I've seen it in other codebases.

**Comment by paulbutcher** [comment]
> This isn't terribly difficult to go through and fix, but it results in ugly redundant imports:

Forgive me, I suspect I'm missing something, but I'm not sure I understand why being able to reopen the namespace allows you to avoid those ugly redundant imports? Can you help me understand?

**Comment by magnus-madsen** [comment]
@paulbutcher 

- For this "Declarations in a parent namespace should not be visible in a sub namespace." I think you are right that we already have it (or never have it, if you will).

- For this "It is illegal to shadow a declaration or type with a use or import." and this " It is illegal to shadow a use or import with a declaration or type." it seems we also have it (pending more testing), but at least the error messages are not great. We use the same for duplicate declarations, but an import/use is not a declaration, so here I would like to introduce specific error messages in these cases.

**Comment by magnus-madsen** [comment]
> With the current implementation I've always assumed you could extend the stdlib in local code:

This was the originally intention-- which we are now discussing whether to backtrack on.

**Comment by mlutze** [comment]
> > With the current implementation I've always assumed you could extend the stdlib in local code:
> 
> This was the originally intention-- which we are now discussing whether to backtrack on.

I am strongly opposed allowing extension of foreign namespaces. The reason is that if two library authors both happen to want to add `List.revAppend`, for example, the libraries are immediately incompatible.

**Comment by magnus-madsen** [comment]
Related, but to be discussed separately: https://github.com/flix/flix/issues/4800

**Comment by magnus-madsen** [comment]
By the way, am I the only one who is confused by our use of `/` and `.`?:

```
use Fixpoint.SubstitutePredSym
use Fixpoint/SubstitutePredSym.substitute
```

**Comment by mlutze** [comment]
> By the way, am I the only one who is confused by our use of `/` and `.`?:
> 
> ```
> use Fixpoint.SubstitutePredSym
> use Fixpoint/SubstitutePredSym.substitute
> ```

This is a consistent source of frustration for everyone I think :)

**Comment by paulbutcher** [comment]
> * For this "It is illegal to shadow a declaration or type with a use or import." and this " It is illegal to shadow a use or import with a declaration or type." it seems we also have it (pending more testing), but at least the error messages are not great. We use the same for duplicate declarations, but an import/use is not a declaration, so here I would like to introduce specific error messages in these cases.

This is the current error message:

```
❌ -- Name Error -------------------------------------------------- file:///Users/paulbutcher/Projects/Flix/flix-playground/src/Main.flix

>> Duplicate declaration 'Foo'.

6 |     use A.Foo
            ^^^^^
            the first occurrence was here.

7 |     enum Foo
             ^^^
             the second occurrence was here.
```

Which seems pretty clear to me? What would you want it changed to?

Regarding:

> (pending more testing)

Here's what we have so far:

https://github.com/flix/flix/blob/master/main/test/ca/uwaterloo/flix/language/phase/TestNamer.scala#L924-L980

Can you let me know what you feel is missing?

**Comment by magnus-madsen** [comment]
The error message is reasonable, but I think now that we have a much clearer design the error messages should reflect that.

I would propose something along the lines:

```
❌ -- Name Error -------------------------------------------------- file:///Users/paulbutcher/Projects/Flix/flix-playground/src/Main.flix

>> Shadowed name 'Foo'. The name 'Foo' cannot be both use'd and declared within the same module.

6 |     use A.Foo
            ^^^^^
            the name 'Foo' is use'd here.

7 |     enum Foo
             ^^^
             the name 'Foo' is declared here.
```

**Comment by magnus-madsen** [comment]
Rust:
![image](https://user-images.githubusercontent.com/2287953/199010417-23aec392-fa9e-472e-9af0-18b836505fea.png)

**Comment by paulbutcher** [comment]
> The error message is reasonable, but I think now that we have a much clearer design the error messages should reflect that.

I see what you're trying to achieve, but I'm concerned. Upper case names can come from six different sources:

* Type classes
* Enums
* Type aliases
* Effects
* Uses
* Imports

Right now, if any of them clash, we generate a `DuplicateUpperName` error.

Are you suggesting that we generate a different error for each (i.e. six different error messages)? Or one for each type of clash (i.e. 30 different error messages)?

Or are you just suggesting that we treat uses and imports differently? But if so, what is it about uses/imports that means that we should treat them differently from other types of name clash?

**Comment by mlutze** [comment]
> > The error message is reasonable, but I think now that we have a much clearer design the error messages should reflect that.
> 
> I see what you're trying to achieve, but I'm concerned. Upper case names can come from six different sources:
> 
>     * Type classes
> 
>     * Enums
> 
>     * Type aliases
> 
>     * Effects
> 
>     * Uses
> 
>     * Imports
> 
> 
> Right now, if any of them clash, we generate a `DuplicateUpperName` error.
> 
> Are you suggesting that we generate a different error for each (i.e. six different error messages)? Or one for each type of clash (i.e. 30 different error messages)?
> 
> Or are you just suggesting that we treat uses and imports differently? But if so, what is it about uses/imports that means that we should treat them differently from other types of name clash?

We could make it so the messages are generated like `you already ${defined/imported/used} your thing here`. But I'm not sure it's a good idea to complicate things so much.

**Comment by paulbutcher** [comment]
I guess we could have `DuplicateUpperName` sniff the types of the symbols it's given and generate different text depending on what they are?

**Comment by magnus-madsen** [comment]
> I guess we could have `DuplicateUpperName` sniff the types of the symbols it's given and generate different text depending on what they are?

I think its more about the source locations: One is the "use" and one is the "re-declaration". That's what should be separated.

**Comment by magnus-madsen** [comment]
`case class ShadowedUpperName(name: String, locUse: SourceLocation, locDecl: SourceLocation)` or so

**Comment by magnus-madsen** [comment]
> OK, so you are suggesting that clashes caused by `use` and `import` are materially different from clashes caused by other means of introducing upper case names? So the current error message is sufficient in this case, for example?:

I think it would be a good starting point. Like, if you write `enum Foo` and `class Foo` hopefully you will see it (with the help of the error message). But often when writing code I never "look at" uses/imports, so I think the error should call attention to it. 

Also to say that such shadowing is explicitly forbidden.

**Comment by paulbutcher** [comment]
OK. So you're suggesting two additional error messages (three in total). One for clashes introduced by a use, one for clashes introduced by an import, and one for everything else?

This honestly makes no sense to me (it strikes me it should either be one error message, or one for each way of introducing a clash). But I'll defer to your judgement.

**Comment by magnus-madsen** [comment]
> OK. So you're suggesting two additional error messages (three in total). One for clashes introduced by a use, one for clashes introduced by an import, and one for everything else?
> 
> This honestly makes no sense to me (it strikes me it should either be one error message, or one for each way of introducing a clash). But I'll defer to your judgement.

Ah, it does seem a bit arbitrary. I think its just trying to be on the path towards more specific error messages. But at the same time not forcing us to have 30 different variants immediately. Does that make sense?

**Comment by paulbutcher** [comment]
The more I think about this, the more I think the error class should be the thing that makes the determination. Consider this example (which is the one that generated the error message above):

```
namespace A {
    pub enum Foo
}

namespace B {
    use A.Foo
    enum Foo
}
```

The point where we detect the clash is when processing the `enum Foo`, not the import. I.e. `Foo` is introduced into the namespace by `use` before it's introduced by `enum`. So the code that detects the clash is the `WeededAst.Declaration.Enum` case within the namer, and it would have to look at the type of the thing that its clashing with to determine which error message to output. The same holds true for every point where a clash might happen

So, it strikes me that we should fix the properly and have `DuplicateUpperName` generate the text of the error message dynamically based on the types of the symbols it's passed?

**Comment by paulbutcher** [comment]
I've been beating my head against this "generate a more specific error message for use and import" issue and I just can't find a good way to do it.

The problem is that the thing that generates the actual errors is `mkDuplicateNamePair` which doesn't have the information to know which type of error to generate.

The code which calls `mkDuplicateNamePair` has *half* of the information necessary (it knows what type of name it's currently dealing with) but has no idea what the source of the clash is.

Then `lookupUpperName` has the *other* half of the information necessary (it knows what the pre-existing name is) but doesn't have the information about what the "new" name is.

Bottom line, I just don't think this is possible without significantly re-structuring the way that we generate these errors.

That's obviously possible, but I'm really unconvinced that it's worth the effort. Let me know what you think.

**Comment by magnus-madsen** [comment]
I want to revisit myself (but a little short on time).

What about merge of uenv and ienv?

**Comment by paulbutcher** [comment]
> What about merge of uenv and ienv?

That isn't done yet, no

**Comment by paulbutcher** [comment]
(so, sorry, that's not in the "Later" section and not yet done)
